misc/peer-store/src/store.rs

```diff
@@ -0,0 +1,99 @@
+use std::{
+    collections::{HashMap, HashSet},
+    time::SystemTime,
+};
+
+use libp2p_core::{Multiaddr, PeerId};
+
+/// A store that
+/// - keep track of currently connected peers;
+/// - contains all observed addresses of peers;
+pub trait Store {
+    /// Called when a peer connects.
+    fn on_peer_connect(&mut self, peer: &PeerId);
+    /// Called when a peer disconnects.
+    fn on_peer_disconnect(&mut self, peer: &PeerId);
+    /// Update an address record.  
+    /// Return `true` when the address is new.  
+    fn on_address_update(&mut self, peer: &PeerId, address: &Multiaddr) -> bool;
+    fn list_connected(&self) -> Box<[&PeerId]>;
+    fn addresses_of_peer(&self, peer: &PeerId) -> Option<Box<[super::AddressRecord]>>;
+}
+
+pub(crate) struct PeerAddressRecord {
+    addresses: HashMap<Multiaddr, AddressRecord>,
+}
+impl PeerAddressRecord {
+    pub(crate) fn records(&self) -> Box<[super::AddressRecord]> {
+        self.addresses
+            .iter()
+            .map(|(address, record)| super::AddressRecord {
+                address,
+                last_seen: &record.last_seen,
+            })
+            .collect()
+    }
+    pub(crate) fn new(address: &Multiaddr) -> Self {
+        let mut address_book = HashMap::new();
+        address_book.insert(address.clone(), AddressRecord::new());
+        Self {
+            addresses: address_book,
+        }
+    }
+    pub(crate) fn on_address_update(&mut self, address: &Multiaddr) -> bool {
+        if let Some(record) = self.addresses.get_mut(address) {
+            record.update_last_seen();
+            false
+        } else {
+            self.addresses.insert(address.clone(), AddressRecord::new());
+            true
+        }
+    }
+}
+
+pub(crate) struct AddressRecord {
+    /// The time when the address is last seen.
+    last_seen: SystemTime,
+}
+impl AddressRecord {
+    pub(crate) fn new() -> Self {
+        Self {
+            last_seen: SystemTime::now(),
+        }
+    }
+    pub(crate) fn update_last_seen(&mut self) {
+        self.last_seen = SystemTime::now();
+    }
+}
+
+/// A in-memory store.
+pub struct MemoryStore {
+    /// Peers that are currently connected.
+    connected_peers: HashSet<PeerId>,

```

Comment:
The idea of `Store` trait is to allow on-disk storage, now I think about it, this info will be changing in real time so it should be kept in memory anyway. Will move it into the behaviour itself.

Reply:
can we have `MemoryStore<T=()>` so that we are able to store data for peers (like scoring etc):

Reply:
> can we have `MemoryStore<T=()>` so that we are able to store data for peers (like scoring etc):

`MemoryStore` is more of a reference implementation, I don't think it is necessary to include a generic parameter for customization since *we* are maintaining its internals.